common/recordCreate.js

```diff
@@ -0,0 +1,300 @@
+             * submissionRowsCopy is the datastructure that will be used for creating
+             * the upload url. It must have all the visible and invisible data.
+             * The following makes sure that submissionRowsCopy has all the underlying data
+             */
+            if (isUpdate) {
+                for (var i = 0; i < submissionRowsCopy.length; i++) {
+                    var newData = submissionRowsCopy[i];
+                    var oldData = $rootScope.tuples[i].data;
+
+                    // make sure submissionRowsCopy has all the data
+                    for (var key in oldData) {
+                        if (key in newData) continue;
+                        newData[key] = oldData[key];
+                    }
+                }
+            }
+
+            //call uploadFiles which will upload files and callback on success
+            uploadFiles(submissionRowsCopy, isUpdate, function() {
+
+                var fn = "create",
+                    args = [submissionRowsCopy];
+                var fnScope = isModalUpdate ? derivedref.unfilteredReference.contextualize.entryCreate : $rootScope.reference.unfilteredReference.contextualize.entryCreate;
+
+                if (isUpdate) {
+
+                    /**
+                     * After uploading files, the returned submissionRowsCopy contains
+                     * new file data. This includes filename, filebyte, and md5.
+                     * The following makes sure that all the data are updated.
+                     * That's why this for loop must be after uploading files and not before.
+                     * And we cannot just pass submissionRowsCopy to update function, because
+                     * update function only accepts array of tuples (and not just key-value pair).
+                     */
+                    for (var i = 0; i < submissionRowsCopy.length; i++) {
+                        var row = submissionRowsCopy[i];
+                        var data = $rootScope.tuples[i].data;
+                        // assign each value from the form to the data object on tuple
+                        for (var key in row) {
+                            data[key] = (row[key] === '' ? null : row[key]);
+                        }
+                    }
+
+                    // submit $rootScope.tuples because we are changing and
+                    // comparing data from the old data set for the tuple with the updated data set from the UI
+                    fn = "update", fnScope = $rootScope.reference, args = [$rootScope.tuples];
+                }
+
+                fnScope[fn].apply(fnScope, args).then(function success(result) {
+
+                    var page = result.successful;
+                    var failedPage = result.failed;
+
+                    // the returned reference is contextualized and we don't need to contextualize it again
+                    var resultsReference = page.reference;
+                    if (isUpdate) {
```

Comment:
This if condition is doing almost the same exact thing as L100-L107. This loop that looks through `submissionRowsCopy` should be factored out as a function that's used in both places.